src/server/pkg/deploy/images/images.go

```diff
@@ -0,0 +1,111 @@
+package images
+
+import (
+	"fmt"
+	"io"
+	"time"
+
+	docker "github.com/fsouza/go-dockerclient"
+	"github.com/pachyderm/pachyderm/src/server/pkg/deploy/assets"
+)
+
+// Export a tarball of the images needed by a deployment.
+func Export(opts *assets.AssetOpts, out io.Writer) error {
+	client, err := docker.NewClientFromEnv()
+	if err != nil {
+		return err
+	}
+	authConfigs, err := docker.NewAuthConfigurationsFromDockerCfg()
+	if err != nil {
+		return fmt.Errorf("error parsing auth: %s, try running `docker login`", err.Error())
+	}
+	if len(authConfigs.Configs) == 0 {
+		return fmt.Errorf("didn't find any valud auth configurations")
+	}
+	images := assets.Images(opts)
+	for _, image := range images {
+		repository, tag := docker.ParseRepositoryTag(image)
+		pulled := false
+		var loopErr error
+		for _, authConfig := range authConfigs.Configs {
+			if err := client.PullImage(
+				docker.PullImageOptions{
+					Repository:        repository,
+					Tag:               tag,
+					InactivityTimeout: 10 * time.Millisecond,
+				},
+				authConfig,
+			); err != nil {
+				loopErr = err
+				continue
+			}
+			pulled = true
+			break
+		}
+		if !pulled {
+			return fmt.Errorf("error pulling image: %+v", loopErr)
+		}
+	}
+	return client.ExportImages(docker.ExportImagesOptions{
+		Names:        images,
+		OutputStream: out,
+	})
+}
+
+// Import a tarball of the images needed by a deployment such as the one
+// created by Export and push those images to the registry specific in opts.
+func Import(opts *assets.AssetOpts, in io.Reader) error {
+	client, err := docker.NewClientFromEnv()
+	if err != nil {
+		return err
+	}
+	authConfigs, err := docker.NewAuthConfigurationsFromDockerCfg()
+	if err != nil {
+		return fmt.Errorf("error parsing auth: %s, try running `docker login`", err.Error())
+	}
+	if len(authConfigs.Configs) == 0 {
+		return fmt.Errorf("didn't find any valid auth configurations")
+	}
+	if err := client.LoadImage(docker.LoadImageOptions{
+		InputStream: in,
+	}); err != nil {
+		return err
+	}
+	registry := opts.Registry
+	opts.Registry = "" // pretend we're using default images so we can get targets to tag
+	images := assets.Images(opts)
+	opts.Registry = registry
+	for _, image := range images {
+		repository, tag := docker.ParseRepositoryTag(image)
+		registryRepo := assets.AddRegistry(opts.Registry, repository)
+		if err := client.TagImage(image, docker.TagImageOptions{
+			Repo: registryRepo,
+			Tag:  tag,
+		},
+		); err != nil {
+			return fmt.Errorf("error tagging image: %v", err)
+		}
+		pushed := false
+		var loopErr error
+		for _, authConfig := range authConfigs.Configs {
```

Comment:
Same thing for pulling the images. As far as I can tell the authconfigs are separate username/passwords? (authenticating across all users seems bad). But if there's more than one for the same person ... presumably both logins won't work (they might be for different registries?) ... in which case this code would always perform the action (push), but since one fails return an error.

Or maybe I'm missing something.

Reply:
So what's going on here is that only one of these auth configs is valid for the registry we're trying to push to. But there doesn't seem to be a great way to figure out which one that is. So the least fragile thing I could think of was just trying all of them. The code is setup in such a way that if pushing to all registries fails then it'll return an error. I think I'm going to make it so it returns all errors so the user has a better chance to diagnose the error.

Reply:
I see. Yea, I think this is fine